### PR TITLE
build: group webpack & ignore dist & support yarn1.x

### PR DESCRIPTION
**Pull request by curly210102**
## :bookmark_tabs: Summary
@Yash-Singh1  Ref: #2456 
- [x] Bulid
   - [x] Don't push dist to GitHub
   - [x] Group webpack files

## :straight_ruler: Design Decisions
1. Group webpack configs to `.webpack` folder, and share base config with `webpack-merge`.
2. Move html files in `dist` directory to `demo` directory
3. improve husky + lint-stage, change `yarn dlx` to `yarn run`, as `dlx` only supported on version 2.x

### :clipboard: Tasks
Make sure you
- [x] :book: have read the [contribution guidelines](https://github.com/mermaid-js/mermaid/blob/develop/CONTRIBUTING.md) 
- [ ] :computer: have added unit/e2e tests (if appropriate) 
- [x] :bookmark: targeted `develop` branch

### CODE REVIEW

File: .eslintignore [hunk 1]
```diff
@@ -1,2 +1,4 @@
 dist/**
-.github/**
+yarn.lock
+demos/**
```
Comment (Yash-Singh1) [inline .eslintignore:3]:
```suggestion
demos/**
!demos/*.html
```

This will make the `eslintignore` compatible with #2531

Comment (curly210102) [reply .eslintignore:3]: Maybe remove `demos/**` from `eslintignore? Only html files in demos directory.

Comment (Yash-Singh1) [reply .eslintignore:3]: Sure! 👍🏾

File: .eslintignore [hunk 2]
```diff
@@ -1,2 +1,4 @@
...
+**/*.json
```
Comment (mmorel-35) [inline .eslintignore:4]: Why are you ignoring json files ?
```suggestion
```

Comment (curly210102) [reply .eslintignore]: eslint+lint-staged think package.json is invalid and refuse commit.
Screen Shot 2021-12-10 at 8 07 36 PM

Comment (mmorel-35) [reply .eslintignore]: Then let's just ignore package.json and not all the json files

Comment (curly210102) [reply .eslintignore]: In fact, this happens on all json files more than just package.json. For example, I changed cypress.json and add it to stage and exec `yarn lint-staged`
Screen Shot 2021-12-10 at 8 17 14 PM

But it doesn't throw error when just exec `yarn lint`

Comment (mmorel-35) [reply .eslintignore]: What is causing this the lint-staged configuration or the eslint configuration?

Comment (curly210102) [reply .eslintignore]: In my understanding, ESLint is a javascript linter which is used to lint js/ts/jsx. When exec `yarn eslint .`, which means lint all js files in project. But when ESLint is used with lint-staged, lint-staged will provides it with other types of files, the parser '@babel/eslint-parser' can not work with json.

And I noticed that jsx was enabled in parserOptions, although I didn't find anywhere to use it. Maybe set .lintstagedrc.json with

``` json
{
  "*.{js,jsx}": ["yarn lint:fix", "git add"],
  "*.{json,md,css,scss,html,yml}": ["prettier --write", "git add"]
}
```

Comment (curly210102) [reply .eslintignore]: lint staged files:
- js and jsx: lint by ESLint
- json,md,css,scss,html,yml: format by prettier

**Comment by knsv** [comment]
is this ready to be merged?

**Comment by mmorel-35** [comment]
I have handled everything in #2551 , thank you @curly210102 for your great contribution

**Pull request closed without merging**